gensim/utils.py

```diff
@@ -1146,27 +1146,33 @@ def keep_vocab_item(word, count, min_count, trim_rule=None):
         else:
             return default_res
 
-def check_output(stdout=subprocess.PIPE, *popenargs, **kwargs):
-    r"""Run command with arguments and return its output as a byte string.
-    Backported from Python 2.7 as it's implemented as pure python on stdlib.
-    >>> check_output(args=['/usr/bin/python', '--version'])
-    Python 2.6.2
-    Added extra KeyboardInterrupt handling
+def check_output(args, flag=True):
+    r"""
+    subprocess.check_output with the flag set to true will spawn a new shell process and execute 'args'
+    if there is an error while executing args, the error message will be displayed on stdout along with the
+    custom error message in the except block. This allows the user to receive an accurate error message if subprocess
+    fails to execute the specified command.
+    If flag is set to true, subprocess.check_output takes 'args' as a string instead of a list. To abstract the user from this,
+    this function will convert the argument list to a string if needed.
+    >>> test_checkoutput(args=['/usr/bin/python', '--version'])
+    Python 2.7.10
+    In case args generates an error
+    >>> test_checkoutput(args=['/usr/bin/pythons', '-ve']) #Incorrect argument
+    /bin/sh: /usr/bin/pythons: command not found
+    subprocess.check_output could not execute command ' /usr/bin/pythons -ve '
     """
+    if flag:
+        args = " ".join(args)
     try:
-        process = subprocess.Popen(stdout=stdout, *popenargs, **kwargs)
-        output, unused_err = process.communicate()
-        retcode = process.poll()
-        if retcode:
-            cmd = kwargs.get("args")
-            if cmd is None:
-                cmd = popenargs[0]
-            error = subprocess.CalledProcessError(retcode, cmd)
-            error.output = output
-            raise error
-        return output
-    except KeyboardInterrupt:
-        process.terminate()
+        res = subprocess.check_output(args, shell=flag)
+        return res
+    except subprocess.CalledProcessError:
+        """
+        If this error is raised, it is because check_output could not execute the command.
+        Instead of raising the error, output a more specific error message
+        """
+        error = "subprocess.check_output could not execute command ' " + str(args) + " '"
+        logger.error(error)
         raise
```

Comment:
I may be missing something, but why exactly do we need to raise the ```CalledProcessError``` at all. As long as we tell the user why ```check_output``` failed, shouldn't that be fine? Using ```shell=true``` with ```subprocess.check_output```, in case of erroneous input, the same error message gets displayed as it would if the user were to enter the command on the terminal. This serves the purpose of telling the user why ```check_output``` failed.

If I pass ```ldfs``` to ```check_output```, stderr will have ```/bin/sh: ldfs: command not found```. Similarly, if you were to type the same command on the terminal, the output would be ```bash: ldfs: command not found```.

In the earlier code stderr would have had a huge error message indicating a ```FileNotFound```. If required ```CalledProcessError``` can be captured and a similar error message can be logged.

In case I've missed something and there is some reason that an implementation similar to the earlier one is preferred, I will look into that.